### PR TITLE
Add function build_arg_list for building arguments list from keyword dictionaries

### PR DESCRIPTION
**Pull request by seisman**
This PR contains the changes of commit 1149b3719a86465308e79d6b8b0835e3dc524137 in PR #3132.

In this PR, a new utility function `build_arg_list` function is added. The new function is modified from the `build_arg_string` function and it differs from the `build_arg_string` function in the following aspects:

1. returns a list of arguments, instead of a single argument string
2. no hacky solutions for handling whitespaces
3. Only apply the `non_ascii_to_octal` function to the single-letter option arguments, not to the configuration arguments and input/output files
4. Doctests are also simplified.

The goal is to replace `build_arg_string` with `build_arg_list` in all module wrappers and deprecate/remove `build_arg_string` in future releases.

PR #3132 contains more changes that we will apply after this PR is merged.

### CODE REVIEW

File: pygmt/helpers/utils.py [hunk 1]
```diff
@@ -315,6 +315,108 @@ def non_ascii_to_octal(argstr):
...
+    >>> print(build_arg_list(dict(R="1/2/3/4", J="X4i", watre=True)))
+    Traceback (most recent call last):
+      ...
+    pygmt.exceptions.GMTInvalidInput: Unrecognized parameter 'watre'.
+    """
+    gmt_args = []
+    for key, value in kwdict.items():
+        if len(key) > 2:  # Raise an exception for unrecognized options
+            raise GMTInvalidInput(f"Unrecognized parameter '{key}'.")
+        if value is None or value is False:  # Exclude arguments that are None and False
```
Comment (yvonnefroehlich) [inline pygmt/helpers/utils.py:397]:
```suggestion
        if value is None or value is False:  # Exclude arguments that are None or False
```

File: pygmt/helpers/utils.py [hunk 2]
```diff
@@ -315,6 +315,108 @@ def non_ascii_to_octal(argstr):
...
+) -> list[str]:
+    r"""
+    Convert keyword dictionaries and input/output files into a list of GMT arguments.
+
+    Make sure all values in ``kwdict`` have been previously converted to a string
+    representation using the ``kwargs_to_strings`` decorator. The only exceptions are
+    ``True``, ``False`` and ``None``.
+
+    Any lists or tuples left will be interpreted as multiple entries for the same
+    command line option. For example, the kwargs entry ``"B": ["xa", "yaf"]`` will be
```
Comment (weiji14) [inline pygmt/helpers/utils.py:332]: Call it 'command-line option' or 'parameter'? Xref https://github.com/GenericMappingTools/pygmt/issues/886.

```suggestion
    Any remaining lists or tuples will be interpreted as multiple entries for the same
    parameter. For example, the kwargs entry ``"B": ["xa", "yaf"]`` will be
```

File: pygmt/helpers/utils.py [hunk 3]
```diff
@@ -315,6 +315,108 @@ def non_ascii_to_octal(argstr):
     return argstr.translate(str.maketrans(mapping))
 
 
+def build_arg_list(
+    kwdict: dict,
+    confdict: dict | None = None,
+    infile: str | pathlib.PurePath | list[str | pathlib.PurePath] | None = None,
```
Comment (weiji14) [inline pygmt/helpers/utils.py:321]: Either a list or tuple can be passed in, so maybe use [`Sequence`](https://docs.python.org/3/glossary.html#term-sequence)? Or `list | tuple`?
```suggestion
    infile: str | pathlib.PurePath | Sequence[str | pathlib.PurePath] | None = None,
```

File: pygmt/helpers/utils.py [hunk 4]
```diff
@@ -315,6 +315,108 @@ def non_ascii_to_octal(argstr):
     return argstr.translate(str.maketrans(mapping))
 
 
+def build_arg_list(
+    kwdict: dict,
```
Comment (weiji14) [inline pygmt/helpers/utils.py:319]: Mention what's in the `dict`?
```suggestion
    kwdict: dict[str, Any],
```

File: pygmt/helpers/utils.py [hunk 5]
```diff
@@ -315,6 +315,108 @@ def non_ascii_to_octal(argstr):
...
+    confdict: dict | None = None,
```
Comment (weiji14) [inline pygmt/helpers/utils.py:320]:
```suggestion
    confdict: dict[str, str] | None = None,
```

**Review by weiji14 — COMMENTED** [review]
Code logic ok, just some suggestions related to the type hints.

**Review by weiji14 — APPROVED** [review]

**Pull request merged**